Upgrades Spring to 5.3.18 and Spring Boot to 2.5.12

zregvart opened the pull request:
Upgrading Spring alone would not have been an issue, upgrading Spring Boot seems to have been a bigger issue. One issue unrelated to Spring upgrades is that the WSDL stack we use seems to fetch and parse remote XML schemas without following HTTP redirects, so I've added local cache for those.

Here are the results of that:

[fix(extension): don't use removed method](https://github.com/syndesisio/syndesis/commit/c51c578ff24f173d4ada318becb54450e5fb491d)

The method `Layout::getLibraryDestination` was renamed to `Layout::getLibraryLocation`.

[fix(api-generator): don't load remote schemas](https://github.com/syndesisio/syndesis/commit/800357b3a2dd7148acf92f3429e23bc029cbc18c)

We should not reach out and load remote XML schemas from
schemas.xmlsoap.org.

Background: it seems that recently schemas.xmlsoap.org website
configured HTTP to HTTPS redirect and in doing so broke our WSDL parser
as it will try to fetch without following HTTP redirects. This results
in empty HTTP body that the parser tries to parse and fails with:

```
org.xml.sax.SAXParseException: Premature end of file.
```

To resolve this this stores the two identified XSD files locally and
sets the custom EntityResolver via a custom WSDLLocator to load those
instead. The same is done for the CXF schema parsing classes which can
load the catalog files from `META-INF/jax-ws-catalog.xml`.

[fix(deps): missing javax.annotation.Nullable](https://github.com/syndesisio/syndesis/commit/e3180d7084e6e694f0274308236e464b56d45622)


Seems that some dependency was pulling this in, but it is not pulling it in any more after Spring/Spring Boot upgrade. This adds it to prevent a compile error.

[fix(extension): compile issues Spring Boot upgrade](https://github.com/syndesisio/syndesis/commit/e836abe3ba3893ffb376caffab118a06b771e9d9)

The `AbstractDependencyFilterMojo:filterDependencies` method that we override has become `final`, so this copies and repackages `AbstractDependencyFilterMojo` and all of the code that is referenced from it in a package private/protected way so we can compile the plugin. 

The Spring Boot code included here is ASF licensed, so it should be okay.

Some code quality/style checks changes were required.

[fix(deps): upgrade to Spring Cloud 2020.0.4](https://github.com/syndesisio/syndesis/commit/94722040e53326c22b09fff835af2f95bf61b84b)

Upgrades to the next version after Hoxton.SR12, as Hoxton.SR12 was not compatible with Spring Boot 2.5.x.

[fix(deps): declare hystrix-core version](https://github.com/syndesisio/syndesis/commit/b208a2f26ecfcf4341f35cba386f2d25cb7a5910)

I think Spring Cloud was declaring it and then dropped it from the release we need to upgrade to, so we need to declare it here instead.

[fix(deps): add missing dependency verisions](https://github.com/syndesisio/syndesis/commit/00175b065bd84e73ef2b779414c93e7277b763fd)

Upgrading to Spring Cloud we lost some of the versions declared here, I picked to the best of my ability the versions, there could be a better selection.

[fix(server): enable prometheus export](https://github.com/syndesisio/syndesis/commit/21dba005288794b7d48ee88d7aaaa2bff5ba8b7a)

It is no longer by default on if Prometheus exporter classes are detected on the classpath, we need to enable it explicitly.

Also adds the `debug=true` to the to the server integration tests so we can tell why a bean wasn't introduced into the context.

[fix(server): adapt KubernetesClient test](https://github.com/syndesisio/syndesis/commit/cf843b2a147650d682de55e509e3cefc238dbba9)

Seems that we need to explicitly load the `Application` configuration, otherwise Spring Boot auto-configuration doesn't kick in.

dependabot[bot] committed:
chore(deps): bump spring-boot-starter-web in /app

Bumps [spring-boot-starter-web](https://github.com/spring-projects/spring-boot) from 2.3.12.RELEASE to 2.5.12.
- [Release notes](https://github.com/spring-projects/spring-boot/releases)
- [Commits](https://github.com/spring-projects/spring-boot/compare/v2.3.12.RELEASE...v2.5.12)

---
updated-dependencies:
- dependency-name: org.springframework.boot:spring-boot-starter-web
  dependency-type: direct:production
...

dependabot[bot] committed:
chore(deps): bump spring-core from 5.3.15 to 5.3.18 in /app

Bumps [spring-core](https://github.com/spring-projects/spring-framework) from 5.3.15 to 5.3.18.
- [Release notes](https://github.com/spring-projects/spring-framework/releases)
- [Commits](https://github.com/spring-projects/spring-framework/compare/v5.3.15...v5.3.18)

---
updated-dependencies:
- dependency-name: org.springframework:spring-core
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

zregvart committed:
fix(extension): don't use removed method

The method `Layout::getLibraryDestination` was renamed to
`Layout::getLibraryLocation`.

zregvart committed:
fix(deps): missing javax.annotation.Nullable

Seems that some dependency was pulling this in, but it is not pulling it
in any more after Spring/Spring Boot upgrade. This adds it to prevent a
compile error.

zregvart committed:
fix(extension): compile issues Spring Boot upgrade

The `AbstractDependencyFilterMojo:filterDependencies` method that we
override has become `final`, so this copies and repackages
`AbstractDependencyFilterMojo` and all of the code that is referenced
from it in a package private/protected way so we can compile the plugin.

The Spring Boot code included here is ASF licensed, so it should be
okay.

Some code quality/style checks changes were required.

zregvart committed:
fix(api-generator): don't load remote schemas

We should not reach out and load remote XML schemas from
schemas.xmlsoap.org.

Background: it seems that recently schemas.xmlsoap.org website
configured HTTP to HTTPS redirect and in doing so broke our WSDL parser
as it will try to fetch without following HTTP redirects. This results
in empty HTTP body that the parser tries to parse and fails with:

```
org.xml.sax.SAXParseException: Premature end of file.
```

To resolve this this stores the two identified XSD files locally and
sets the custom EntityResolver via a custom WSDLLocator to load those
instead. The same is done for the CXF schema parsing classes which can
load the catalog files from `META-INF/jax-ws-catalog.xml`.

zregvart committed:
fix(deps): declare hystrix-core version

I think Spring Cloud was declaring it and then dropped it from the
release we need to upgrade to, so we need to declare it here instead.

zregvart committed:
fix(deps): add missing dependency verisions

Upgrading to Spring Cloud we lost some of the versions declared here, I
picked to the best of my ability the versions, there could be a better
selection.

zregvart committed:
fix(deps): upgrade to Spring Cloud 2020.0.4

Upgrades to the next version after Hoxton.SR12, as Hoxton.SR12 was not
compatible with Spring Boot 2.5.x.

zregvart committed:
fix(server): enable prometheus export

It is no longer by default on if Prometheus exporter classes are
detected on the classpath, we need to enable it explicitly.

Also adds the `debug=true` to the to the server integration tests so we
can tell why a bean wasn't introduced into the context.

zregvart committed:
fix(server): adapt KubernetesClient test

Seems that we need to explicitly load the `Application` configuration,
otherwise Spring Boot auto-configuration doesn't kick in.

phantomjinx reviewed (approved):
LGTM

zregvart committed:
fix(bom): use Spring Boot 2.5.12 in intg and ext

Missed in the commit for upgrading Spring Boot version, we need these
two also aligned.

zregvart committed:
fix(extension): align AtlasMap version

zregvart committed:
fix(deps): force declared Spring Framework version

We can get different version of Spring Framework via Camel dependencies,
e.g. via `camel-spring`, even though we wish to be aligned with Camel
there are situations we might want a newer version (but compatible) of
Spring. This forces the declared version of Spring Framework by
importing the spring-framework-bom with the same version before the
Camel BOM, and that way solidifying the version of Spring Framework to
be used.

zregvart committed:
fix(deps): correct Spring Boot version

Spring projects don't use `.RELEASE` any more.

zregvart commented:
retest this please

mkralik3 commented:
The wrong qe jenkins was triggered, the running job for the latest changes is here: https://fuse-7x-jenkins-csb-fuse-qe.apps.ocp4.prod.psi.redhat.com/view/Fuse%20Online%207.11/job/fuse-7.11-syndesis-tests-PR-ENG/154/console

claudio4j commented:
Add `org.springframework.boot.actuate.autoconfigure.metrics.mongo.MongoMetricsAutoConfiguration` to `spring.autoconfigure.exclude` in `app/meta/src/main/resources/application.yml`, otherwise it throws IllegalStateException: Failed to introspect Class [org.springframework.boot.actuate.autoconfigure.metrics.mongo.MongoMetricsAutoConfiguration$ and a ClassNotFoundException

zregvart committed:
fix(meta): enable Prometheus metrics export

zregvart committed:
fix(meta): exclude unused auto-configurations

We don't need these, and they might fail the application startup as some
of their dependencies might be excluded or unavailable for other
reasons.

zregvart committed:
chore(meta): integration tests

Given that we don't have any integration tests we can't be alerted to
failures of the syndesis-meta configuration. In this particular case we
were not aware that a MongoDB auto-configuration was enabled to be
loaded as the MongoDB classes are on the classpath but failed to load as
a dependency of theirs was not present on the classpath.

```
java.lang.IllegalStateException: Failed to introspect Class [org.springframework.boot.actuate.autoconfigure.metrics.mongo.MongoMetricsAutoConfiguration$MongoConnectionPoolMetricsConfiguration] from ClassLoader [org.springframework.boot
.loader.LaunchedURLClassLoader@4459eb14]

ClassNotFoundException: io.micrometer.core.instrument.binder.mongodb.MongoConnectionPoolTagsProvider
```

zregvart commented:
@claudio4j I've added those exclusions but also a bit more so we're not bitten with this in the future. I've added tests to check if Meta starts up and one test for the `/connectors` endpoint. I've found the same issue as in 21dba005 in Meta as well.

mkralik3 commented:
retest this please

zregvart commented:
retest this please

mkralik3 commented:
The QE job doesn't work properly. We need to fix checking images on Quay. I am working on it.

mkralik3 commented:
retest this please
